### PR TITLE
Redesign PG loading from EEPROM

### PR DESCRIPTION
**Pull request by ledvinap**
All parameter groups  initialized, either from EEPROM or reset to
default.
Stored record version is checked, mismatched version is ignored

### CODE REVIEW

**Comment by ledvinap** [comment]
This should help with #2395.
Review/comments are welcome, but the code is not yet ready for merge. I'll force-push new version, don't base new work on this PR

**Comment by ledvinap** [comment]
@hydra: Is it OK to `config_eeprom.c` functions to start with `EEPROM`?

**Comment by ledvinap** [comment]
Unittests will be fixed soon

**Comment by hydra** [comment]
@ledvinap nice to see some progress on this, cool.

there are a few cases that need to be handled:

1) load existing pg
2) initialise pg's that cannot be found
3) load pg with matching version that has NEW fields at the end of the record - the new fields must be reset to defaults.
4) re-initialise pg's the have a version mismatch.

**Comment by ledvinap** [comment]
> 1) load existing pg

Stored configuration is used when it does exist (correct profile/system +PGN) and version does match

> 2) initialise pg's that cannot be found

PG is reset when EEPROM record is not found

> 3) load pg with matching version that has NEW fields at the end of the record - the new fields must be reset to defaults - size mismatch.

PG is reset first, than stored data are copied over it. 

> 4) re-initialise pg's the have a version mismatch.

Record is reset on version mismatch (nothing is copied over default values)

Handling of arrays may be improved - pad/trim on per-item basis, then reset remaining items if necessary. But that can go into next PR.

Also it will be quite easy to implement PG upgrade mechanism - register function that will chane record to higher version number. Multiple upgrade steps may be performed automatically, oldest supported CF version may be handled by #define, including only upgrade functions that fit available space

**Comment by hydra** [comment]
@ledvinap great. so all the cases outlined above are all handled by this PR as-is?

File: src/main/config/config_eeprom.c [hunk 1]
```diff
@@ -99,35 +100,8 @@ static uint8_t updateChecksum(uint8_t chk, const void *data, uint32_t length)
...
-
-    pgLoad(reg, record->pg, record->size, profileIndex);
-    return true;
-}
-
-// Scan the EEPROM config.  Optionally also load into memory.  Returns
-// true if the config is valid.
-bool scanEEPROM(bool andLoad) // FIXME boolean argument.
+// Scan the EEPROM config. Returns true if the config is valid.
+bool scanEEPROM(void)
```
Comment (martinbudden) [inline src/main/config/config_eeprom.c]: Good to see the boolean parameter gone.

File: src/main/config/config_eeprom.c [hunk 2]
```diff
@@ -167,9 +138,58 @@ bool scanEEPROM(bool andLoad) // FIXME boolean argument.
...
 }
 
+// find config record for reg + classification (profile info) in EEPROM
+// return NULL when record is not found
+// this function assumes that EEPROM content is valid
+const configRecord_t *findEEPROM(const pgRegistry_t *reg, configRecordFlags_e classification)
+{
+    const uint8_t *p = &__config_start;
+    p += sizeof(configHeader_t);             // skip header
+    while(true) {
```
Comment (martinbudden) [inline src/main/config/config_eeprom.c]: ```while (true)``` space after while. Similarly for a few `if`s in the code.

Comment (ledvinap) [inline src/main/config/config_eeprom.c]: Thanks, I'll fix it

**Comment by ledvinap** [comment]
@hydra : I hope so .. It needs testing, I'll prepare some unittests ...

File: src/main/config/config_eeprom.c (hunk 1 above)
Comment (hydra) [inline src/main/config/config_eeprom.c]: woohoo!

**Comment by hydra** [comment]
```@hydra: Is it OK to config_eeprom.c functions to start with EEPROM? ```
if you mean, is it ok to change the prefix of the eeprom functions to have eeprom at the start. yes. but not in this PR please.  as @martinbudden metioned recently let's keep larger renames a little more focussed.

File: src/main/config/config_eeprom.c (hunk 2 above)
Comment (ledvinap) [inline src/main/config/config_eeprom.c]: BTW: find . -type f -exec grep -nH -E '\<(if|switch|for|do|while)[(]' {} +

**Comment by hydra** [comment]
@ledvinap ready?

**Comment by hydra** [comment]
@ledvinap in doing a bit of testing and monitoring the reg->pgn from inside `pgLoad` I noticed unexpected values, such as `4123` and `4107`.  I had basically flashed the 1.13.0 binary, let it boot and then upgraded the firmware to master via SWD to see what what was being reset and what wasn't when i noticed these unexpected values.

**Comment by hydra** [comment]
ahh wait, never mind, lol, that's the version in the top 4 bits.

**Comment by hydra** [comment]
I do note, that after the config is loaded and some settings are reset the config is not immediately saved and the 'upgrade' process is followed on every boot until the config is saved once.

I think that's probably acceptable but I'd like your thoughts.

**Comment by ledvinap** [comment]
IMO repeated upgrade is fine. If you downgrade/change version before saving, you get original config back. Can be especially useful if you flash wrong version.
If config upgrade using intermediate versions is desired, you can save explicitly.

**Comment by ledvinap** [comment]
BTW: It should be quite easy to skip save when config did not change (use code identical to load and compare EEPROM records to PG).

**Comment by hydra** [comment]
yeah, not-saving is useful feature actually :smile:  

nice.

**Comment by ledvinap** [comment]
This PR is not that large, I don't see any problem. But I did not test it on actual hardware.
If you tested it on HW (any board should do), is should be safe to merge.,

**Comment by hydra** [comment]
Yes I tested it on an SP Racing F3 EVO and went from 1.13.0 to the latest master + this PR.

I'll merge it now, we can fix in next RC if something comes up.

**Comment by hydra** [comment]
@ledvinap I'm finding if I upgrade from 1.13.0 to master then cli dump fails due to incorrect value of telemetry_send_cells.  This needs investigation.

**Comment by ledvinap** [comment]
fixed in #2421. The bug was there before this PR, I totally missed it ... At least new code is a bit nicer ...

**Comment by ledvinap** [comment]
Some logs I collected are here:
https://drive.google.com/file/d/0B9HM4UOKDQBLTzVfVDM0Q3EzX1k/view?usp=sharing
It's mix of logs from different users, some may be experimental/damaged

**Pull request merged**